solr/core/src/java/org/apache/solr/handler/component/UBIComponent.java

```diff
@@ -0,0 +1,424 @@
+              + " must include the 'ubiQuery()' to record UBI queries.");
+    }
+
+    streamFactory = new DefaultStreamFactory();
+    streamFactory.withFunctionName("logging", LoggingStream.class);
+    streamFactory.withFunctionName("ubiQuery", UBIQueryStream.class);
+
+    if (coreContainer.isZooKeeperAware()) {
+      String defaultZkHost = core.getCoreContainer().getZkController().getZkServerAddress();
+      streamFactory.withDefaultZkHost(defaultZkHost);
+    }
+  }
+
+  @Override
+  public void prepare(ResponseBuilder rb) throws IOException {
+    SolrParams params = rb.req.getParams();
+    if (!params.getBool(COMPONENT_NAME, false)) {
+      return;
+    }
+    rb.setNeedDocList(true);
+  }
+
+  @Override
+  public void process(ResponseBuilder rb) throws IOException {
+    SolrParams params = rb.req.getParams();
+    if (!params.getBool(COMPONENT_NAME, false)) {
+      return;
+    }
+
+    doStuff(rb);
+  }
+
+  @Override
+  public int distributedProcess(ResponseBuilder rb) throws IOException {
+
+    SolrParams params = rb.req.getParams();
+    if (!params.getBool(COMPONENT_NAME, false)) {
+      return ResponseBuilder.STAGE_DONE;
+    }
+
+    if (rb.stage != ResponseBuilder.STAGE_GET_FIELDS) {
+      return ResponseBuilder.STAGE_DONE;
+    }
+
+    doStuff(rb);
+
+    return ResponseBuilder.STAGE_DONE;
+  }
+
+  public void doStuff(ResponseBuilder rb) throws IOException {
+
+    // not sure why but sometimes we get it twoice...  how can a response have the
+    // the same component run twice?
+    if (rb.rsp.getValues().get("ubi") != null) {
+      return;
+    }
+    SolrParams params = rb.req.getParams();
+
+    SolrIndexSearcher searcher = rb.req.getSearcher();
+
+    String queryId = params.get(QUERY_ID);
+    UBIQuery ubiQuery = new UBIQuery(queryId);
+
+    ubiQuery.setUserQuery(params.get(USER_QUERY));
+    ubiQuery.setApplication(params.get(APPLICATION));
+    if (ubiQuery.getApplication() == null) {
+      ubiQuery.setApplication(
+          rb.isDistrib
+              ? rb.req.getCloudDescriptor().getCollectionName()
+              : searcher.getCore().getName());
+    }
+
+    String queryAttributes = params.get(QUERY_ATTRIBUTES);
+
+    if (queryAttributes != null && queryAttributes.toString().startsWith("{")) {
+      // Look up the original nested JSON format, typically passed in
+      // via the JSON formatted query.
+      @SuppressWarnings("rawtypes")
+      Map jsonProperties = rb.req.getJSON();
+      if (jsonProperties.containsKey("params")) {
+        @SuppressWarnings("rawtypes")
+        Map paramsProperties = (Map) jsonProperties.get("params");
+        if (paramsProperties.containsKey(QUERY_ATTRIBUTES)) {
+          @SuppressWarnings("rawtypes")
+          Map queryAttributesAsMap = (Map) paramsProperties.get(QUERY_ATTRIBUTES);
+          ubiQuery.setQueryAttributes(queryAttributesAsMap);
+        }
+      }
+    }
+
+    ResultContext rc = (ResultContext) rb.rsp.getResponse();
```

Comment:
To make it distributed, we need to split this method here, and execute parts per certain stage.
It seems, UBIComp needs to submit docids found. Right? 
In distributed context, we need to record before-merge per-shards ids or resulting merged&cropped result ids? I bet the later, please confirm.

Reply:
Thanks @mkhludnev  for looking at this, gives me renewed energy to know someone else is looking at it!  Yes, it is the merged&cropped result ids.   We want to record in the plugin what the candidate result ids that the user MIGHT have seen, which later is used to compare against clicks and impressions to identify which docs are NOT attractive to the user.